File: 프로그래머스/2/49993. 스킬트리/스킬트리.py
def solution(skill, skill_trees):
    answer = 0
    skill_sequence = []
    
    for s in range(len(skill) - 1, -1, -1):
        skill_sequence.append(skill[s])

    for i in range(len(skill_trees)):
        can = True
        # copy_skill = skill_sequence 이렇게 두 변수가 동일한 리스트를 참조하게 됨
        copy_skill = skill_sequence[:]
        for j in skill_trees[i]:
            if j in copy_skill:
                if j != copy_skill.pop():
                    can = False
                    break
        if can:
            answer += 1
        
    
    return answer

File: 프로그래머스/2/49993. 스킬트리/스킬트리.java
import java.util.*;

class Solution {
    public int solution(String skill, String[] skill_trees) {
        Set<Character> possibleSkills = new HashSet<>();
        Map<Character, Character> preSkill = new HashMap<>();
        for (char c = 'A'; c <= 'Z'; c++)
            possibleSkills.add(c);
        for (int i = 1; i < skill.length(); i++) {
            possibleSkills.remove(skill.charAt(i));
            preSkill.put(skill.charAt(i-1), skill.charAt(i));
        }
        
        int answer = 0;
        for (String skills : skill_trees) {
            boolean possible = true;
            Set<Character> nowPossibleSkills = new HashSet<>(possibleSkills);
            for (char s : skills.toCharArray()) {
                if (!nowPossibleSkills.contains(s)) {
                    possible = false;
                    break;
                }
                if (preSkill.containsKey(s))
                    nowPossibleSkills.add(preSkill.get(s));
            }
            if (possible){
                answer++;
            }
        }
        
        return answer;
    }
}